pennylane/ops/qubit/state_preparation.py

```diff
@@ -88,6 +86,28 @@ def compute_decomposition(n, wires):
         """
         return [BasisStatePreparation(n, wires)]
 
+    def state_vector(self, wire_order=None):
+        """Returns a state-vector of shape ``(2,) * num_wires``."""
+        state = self.parameters[0]
+        if any(i not in [0, 1] for i in state):
+            raise ValueError("BasisState parameter must consist of 0 or 1 integers.")
+
+        if wire_order is None:
+            num_wires = len(self.wires)
+            indices = state
+        else:
+            new_wires = Wires(wire_order)
+            if not new_wires.contains_wires(self.wires):
+                raise WireError("Custom wire_order must contain all BasisState wires")
+            num_wires = len(new_wires)
+            indices = [0] * num_wires
+            for base_wire_label, value in zip(self.wires, state):
+                indices[new_wires.index(base_wire_label)] = value
+
+        ket = np.zeros((2,) * num_wires)
```

Comment:
I might be misinterpreting the intention behind the creation of this method. Feel free to correct me, but what I'm inferring is that this method will replace the internal `_apply_basis_state` state preparation methods for qubit devices.

If that is the case, this shape will not work for `default.mixed`, where the internal state has shape `(2,) * (2 * num_wires)`.

Reply:
well.. `default.qubit` and `default.mixed` have their own implementations of `_apply_basis_state` 😄 but this is supposed to replace only _part_ of that old function's work, and in my mind it was only for `default.qubit` (or rather, the python simulator that will replace `default.qubit`). But you're right, we should certainly note in the documentation or in our planning that this is not for the mixed device.

The other part of that code's replacement is in #3683, and it's the actual application of the basis state. Pretty minimal rn, but I like it like that 😏

Reply:
Or we can end up adding a keyword argument to toggle between the two options. But we can do that later.

Reply:
have we discussed the plan for migrating `default.mixed` to the new device API in general? Perhaps some of it should be considered before doing toooo much work, in case we want to change some things in our initial plan to account for it even if we aren't immediately implementing that device.

Reply:
Once we have the kinks ironed out, we will gradually move *all* devices over to the new interface.

And we can always have `default.mixed` call this method to get the pure state and then simply convert it to a density matrix.  That way it still relies on this method, but just has an additional function call it needs to make.